teps/0147-dynamic-matrix.md

```diff
@@ -0,0 +1,379 @@
+
+### Goals
+
+- Be able to author a sharable pipeline with `matrix` such that a task can fan out based on the explicit combinations
+which are specified dynamically.
+
+### Non-Goals
+
+- Matrix explicit combination mechanism only supports a `param` of type `string` and not adding support for any
+other type.
+
+### Use Cases
+
+Let's revisit the example of build pipeline from [TEP-0118](0118-matrix-with-explicit-combinations-of-parameters.md#define-explicit-combinations-in-the-matrix).
+A pipeline `pipeline-to-build-images-from-a-single-repo` has three explicit combinations defined in the pipeline such
+that `kaniko` can fan out and create three separate `taskRun`s. The values for each combination can be furthered
+dynamically specified using the `params` as seen in the example in the [summary](#summary).
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-images-from-a-single-repo
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: kaniko-build
+      taskRef:
+        name: kaniko
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      matrix:
+        include:
+          - name: build-1
+            params:
+              - name: IMAGE
+                value: "image-1"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile1"
+          - name: build-2
+            params:
+              - name: IMAGE
+                value: "image-2"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile2"
+          - name: build-3
+            params:
+              - name: IMAGE
+                value: "image-3"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile3"
+    - ...
+```
+
+This `pipeline` is part of a catalog and shared across multiple teams. Now, the teams are building their own applications
+using this `pipeline` and different teams have different number of images to built as part of their application. This way
+of specifying `matrix` combinations in a `pipeline` is constant and can not be utilized for an application with
+different number of images.
+
+## Proposal
+
+We propose adding a field - `strategy` of type `string` - within the `matrix` section. This allows pipeline authors to
+maintain a single pipeline in a catalog and allow the users to specify the list of explicit combinations for each
+instance dynamically through `pipelineRun` or a trigger template or a task result.
+
+The `matrix.strategy` is of type `string` where its value is a list of objects in key:value pairs. The controller reads
+a stringified JSON payload in a form of `{"include": a list of combinations}` and creates an equivalent number of
+instances based on the length of the specified list.
+
+When `matrix.strategy` is specified, no other `matrix` fields will be allowed i.e. `matrix.params` and `matrix.include`
```

Comment:
It sounds like the `strategy` feature is mutually exclusive to the existing { `params`, `include`}. Wondering if this is fully compatible if we are going to put both the existing matrix feature and the new `matrix-strategy` feature under `enable-api-fields` ?

Reply:
yes, `strategy` is mutually exclusive from the existing `params` and `include` from the usage perspective. From the functionality perspective, it is tightly coupled with `include` and follows the similar syntax.

The existing matrix feature is guarded behind `enable-api-fields` and `strategy` will be guarded behind `enable-api-fields` as well.

Is that what you are asking?